pvlib/irradiance.py

```diff
@@ -1198,6 +1198,48 @@ def perez(surface_tilt, surface_azimuth, dhi, dni, dni_extra,
         return sky_diffuse
 
 
+def clearsky_index(ghi, clearsky_ghi, max_clearsky_index=2.0):
+    """
+    Calculate the clearsky index.
+
+    The clearsky index is the ratio of global to clearsky global irradiance.
+    Negative and non-finite clearsky index values will be truncated to zero.
+
+    Parameters
+    ----------
+    ghi : numeric
+        Global horizontal irradiance in W/m^2.
+
+    clearsky_ghi : numeric
+        Modeled clearsky GHI
+
+    max_clearsky_index : numeric, default 2.0
+        Maximum value of the clearsky index. The default, 2.0, allows
+        for over-irradiance events typically seen in sub-hourly data.
+
+    Returns
+    -------
+    clearsky_index : numeric
+        Clearsky index
+    """
+    clearsky_index = ghi / clearsky_ghi
+    # set +inf, -inf, and nans to zero, while preserving input type
+    # but preserve nans in the input arrays as well
+    input_is_nan = ~np.isfinite(ghi) | ~np.isfinite(clearsky_ghi)
+    if isinstance(clearsky_index, pd.Series):
+        clearsky_index[~np.isfinite(clearsky_index)] = 0
+        clearsky_index[input_is_nan] = np.nan
+    else:
+        clearsky_index = np.where(~np.isfinite(clearsky_index), 0,
+                                  clearsky_index)
+        clearsky_index = np.where(input_is_nan, np.nan, clearsky_index)
```

Comment:
since the logic is a little more complicated now, I suggest we remove the `if` statement here and only use `np.where`. Then, right before the return statement, add 

```python
if isinstance(ghi, pd.Series): 
    clearsky_index = pd.Series(clearsky_index, index=ghi.index)
```

This way we separate the type handling from the function logic

Reply:
Good idea, done.